pkg/kubelet/cm/devicemanager/plugin/v1beta1/client.go

```diff
@@ -69,11 +72,22 @@ func (c *client) Connect() error {
 		klog.ErrorS(err, "Unable to connect to device plugin client with socket path", "path", c.socket)
 		return err
 	}
+
+	err = c.handler.PluginConnected(c.resource, c)
+	if err != nil {
+		klog.ErrorS(err, "Failed to connect to device plugin", "resource", c.resource)
+		if err := conn.Close(); err != nil {
+			klog.V(2).ErrorS(err, "Failed to close grcp connection", "resource", c.resource)
+		}
+		return err
+	}
+
```

Comment:
I wonder (and I'm _not_ pushing for this solution, but let's give it a fair evaluation) if we can use a `defer` to catch the issue instead of moving `PluginConnected` above. Something like

```
func (c *client) Connect() (rerr error) {
    // ...

    c.mutex.Lock()
    c.grpc = conn
    c.client = client
    c.mutex.Unlock()
    defer func() {
        if rerr == nil {
            return
        }
        klog.ErrorS(err, "Failed to connect to device plugin", "resource", c.resource)
        if err := conn.Close(); err != nil {
            klog.V(2).ErrorS(err, "Failed to close grcp connection", "resource", c.resource)
           return
        }
        c.grpc = nil
    }()
    return c.handler.PluginConnected(c.resource, c)
}

Reply:
If so, we have to lock again to reset c.grpc and c.client in `defer` func

Reply:
right. On second look the order of operations should not be too critical here. I'll have a deeper look later on.

Reply:
line 80: `Failed to close grcp connection`
"grcp"?

Reply:
fixed. 

And @ffromani I follow your suggestion.